### ISSUE TITLE
Wrong HTTP code is given causing problems for evaluating errors.

### ISSUE THREAD

**Issue by ppostma1** [magento/magento2#20386]
### Preconditions (*)

1. Magento\Customer\Controller\Section\Load.php, use of STATUS_CODE_400 in code
2. Magento\Catalog\Controller\Product\Frontend\Action\Synchronize

### Steps to reproduce (*)
1. Create an error loading a customer section and a 400 error is returned

### Expected result (*)
1. Error code of 500 or greater.

### Actual result (*)

1. PHP _SHOULD_ NEVER return a 400 error code.  Code 400 in the logs indicates that a request was corrupted, invalid, and was NEVER passed on to the backend.
2. However, what is happening is PHP is accepting, processing, and then throwing an HTTP: bad message error itself, falsely indicating that the request was never processed.
3. This happens particularly when people are trying to make a payment with credit card information.
--------------------
[14/Jan/2019:17:23:25 +0000] "POST /rest/default/V1/guest-carts/d9fd9d3dfd211da7a31eb1b270704887/payment-information HTTP/1.1" 400 70 
[14/Jan/2019:17:33:55 +0000] "POST /rest/default/V1/guest-carts/8a5bddda896e3eb9f9d027b57d0f9a2e/payment-information HTTP/1.1" 400 70 
[15/Jan/2019:20:30:08 +0000] "POST /rest/default/V1/guest-carts/7190a89ff96db5d42b8daf02c8be135f/billing-address HTTP/1.1" 400 93 
[15/Jan/2019:20:30:37 +0000] "POST /rest/default/V1/guest-carts/7190a89ff96db5d42b8daf02c8be135f/billing-address HTTP/1.1" 400 93 
[15/Jan/2019:21:46:25 +0000] "POST /rest/default/V1/guest-carts/82373a4a5a58953976166f8f45e4b522/billing-address HTTP/1.1" 400 93 
[16/Jan/2019:13:26:57 +0000] "POST /rest/default/V1/guest-carts/d2eb71d87ffda30bbb50feb3ca9f0df2/payment-information HTTP/1.1" 400 70 
[16/Jan/2019:19:55:48 +0000] "POST /rest/default/V1/guest-carts/18d4aeaf150c1e5f2b8937f219b6cd10/billing-address HTTP/1.1" 400 93 
[16/Jan/2019:19:55:51 +0000] "POST /rest/default/V1/guest-carts/18d4aeaf150c1e5f2b8937f219b6cd10/billing-address HTTP/1.1" 400 93 
[16/Jan/2019:19:56:12 +0000] "POST /rest/default/V1/guest-carts/18d4aeaf150c1e5f2b8937f219b6cd10/billing-address HTTP/1.1" 400 93 
[17/Jan/2019:01:23:40 +0000] "POST /rest/default/V1/guest-carts/fa718511a7aaceda4975f1c2d6cd812e/payment-information HTTP/1.1" 400 79 
[17/Jan/2019:12:46:11 +0000] "POST /rest/default/V1/guest-carts/d050fdaee4298937e44fc30ef109d82c/billing-address HTTP/1.1" 400 93 
[17/Jan/2019:16:12:13 +0000] "POST /rest/default/V1/guest-carts/c01921a30ac8f2db902e56d1a77b1ec9/billing-address HTTP/1.1" 400 93 
-----------

4. This makes it look like people are sending bad data to us and the cards are not being processed.
5. But in fact the credit card is being processed, the attempt to load their customer data failed, and the resulting transaction is lost. Lost without any logs, errors, or exceptions recorded.
6. This HTTP code should be 500, internal server error, and the event should be logged with information. This will permit reviews to catch the server error, and allow developers to retrieve the error information.

**Comment by magento-engcom-team** [comment]
Hi @ppostma1. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@ppostma1 do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by magento-engcom-team** [comment]
Hi @irajneeshgupta. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [x] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [x] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [x] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [x] 6. Add label `Issue: Confirmed` once verification is complete. 

- [x] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by magento-engcom-team** [comment]
@irajneeshgupta Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-97817`, `MAGETWO-97818` were created

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
Looks like this issue is already verified and confirmed. But if your want to validate it one more time, please, go though the following instruction:
- [ ] 1. Add/Edit `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.

- [ ] 2. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 3. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 4. If the issue is not relevant or is not reproducible any more, feel free to close it.

**Comment by ghost** [comment]
Hi @ppostma1 thank you for you report, the error code in Class Load is correct, you always can see response message in log, so there no reason to change error code.

> A 400 Bad Request Error indicates that the server (remote computer) is unable (or refuses) **to process the request** sent by the client (web browser), due to an issue that is perceived by the server to be a client problem. ... If an expected custom HTTP header is missing or invalid, a 400 Bad Request Error is a likely result.

**Comment by ppostma1** [comment]
@engcom-backlog-nazar Except that when this is thrown, it has already been validated and nothing is missing. The exception the 400 is thrown on is not a missing argument or field, but the widest \Exception class, this blankest code errors, database transactions, and the exception is throw after the card is completely transacted, after the quote is converted to an order, but before the order is saved. This then happens loading the customer section... on a guest user.
If the order is being processed, it is no longer a CLIENT-SIDE error. This has something to do with code, database, or settings.
If the exceptions caught were at least `catch(\InvalidArgumentException $iae) { ...\Zend\Http\Response::STATUS_CODE_400 ... } catch(\Exception $d) { ...\Zend\Http\Response::STATUS_CODE_500 ...}`
The response would be better understood in review

**Comment by ppostma1** [comment]
@engcom-backlog-nazar 
```[2019-04-12 07:31:38] main.NOTICE: tripped 400 in Error Processor [] []
[2019-04-12 07:31:38] main.NOTICE: Magento\Framework\Exception\CouldNotSaveException [] []
[2019-04-12 07:31:38] main.NOTICE: Unable to read response, or response is empty [] []
[2019-04-12 07:31:38] main.NOTICE: /usr/share/nginx/html/magento/vendor/magento/module-checkout/Model/GuestPaymentInformationManagement.php:109 [] []
[2019-04-12 07:31:38] main.NOTICE: #0 /usr/share/nginx/html/magento/vendor/magento/framework/Interception/Interceptor.php(58): Magento\Checkout\Model\GuestPaymentInformationManagement->savePaymentInformationAndPlaceOrder(...
```
This is no 400 error. This is an embarrassment to Magento

**Comment by ppostma1** [comment]
The exception thrown by the lower classes to report a 400 error is `InputException`. 
By catching all under a general Exception, this includes Database errors. Which is exactly what it was tracked down to be. If a general exception is caught, and unfiltered, the code should be 500, a general error, because no effort was made to distinguish a user error.

**Comment by ppostma1** [comment]
Thank you!
on Webapi/ErrorProcessor I have built out the following code to handle most of what I have encountered so far:

```

public function maskException(\Exception $exception)
{
	$isDevMode = $this->_appState->getMode() === State::MODE_DEVELOPER;
	$stackTrace = $isDevMode ? $exception->getTraceAsString() : null;

	//Could-not-save wraps a parent exception that contains the details we need to work with.
	if($exception instanceof \Magento\Framework\Exception\CouldNotSaveException){
		$parentException = $exception;
		//not all exceptions are set to contain a previous
		$exception = $exception->getPrevious() ?? $exception;

	}

	if ($exception instanceof WebapiException) {
		$maskedException = $exception;
	} elseif ($exception instanceof LocalizedException) {
		// Map HTTP codes for LocalizedExceptions according to exception type
		if ($exception instanceof NoSuchEntityException) {
			$httpCode = WebapiException::HTTP_NOT_FOUND;
		} elseif (($exception instanceof AuthorizationException)
			|| ($exception instanceof AuthenticationException)) {

			$httpCode = WebapiException::HTTP_UNAUTHORIZED;

		} elseif ($exception instanceof InputException) {
			$httpCode = WebapiException::HTTP_BAD_REQUEST;

		} elseif ($exception instanceof \Magento\Payment\Gateway\Command\CommandException
				|| $exception instanceof \Magento\Payment\Gateway\Http\ClientException) {

				//remote client errors

				if (stripos($exception->getMessage(), 'unable to read response, or response is empty') !== false) {
					//payment or other gateway failed, message from ZendClientException of no_response
					$httpCode = 504; //remote timeout
				} elseif (preg_match('/rejected|refused/i', $exception->getMessage())) {
					//remote or payment did not agree with information given, fix and resend
					$httpCode = 409;
				} else {
					$httpCode = $exception->getCode() ?? WebapiException::HTTP_INTERNAL_ERROR;
					//log this to track it down
					$this->_logger->notice($exception);
				}

		} else {
			// Input, Expired, InvalidState exceptions will fall to here
			$httpCode = $exception->getCode(); // ?? 500;  //Unknown error as default.
			if(empty($httpCode)) {
				if(preg_match('/lost|gone|expired/', $exception->getMessage()))
					$httpCode = 410; //Gone
				else
					$httpCode = WebapiException::HTTP_INTERNAL_ERROR; //default
			}

			$this->_logger->notice("WebAPI Error Processor report");
			$this->_logger->notice($exception); 
			$this->_logger->notice("/end WebAPI Error Processor Report");

		}

		if ($exception instanceof AggregateExceptionInterface
			|| (isset($parentException) && $parentException instanceof AggregateExceptionInterface)) {
			if (isset($parentException) && $parentException instanceof AggregateExceptionInterface)
				//switch back now
				$exception = $parentException;

			$errors = $exception->getErrors();
		} else {
			$errors = null;
		}

		$maskedException = new WebapiException(
			new Phrase($exception->getRawMessage()),
			$exception->getCode(),
			$httpCode,
			$exception->getParameters(),
			get_class($exception),
			$errors,
			$stackTrace
		);
	} else {
		$message = $exception->getMessage();
		$code = $exception->getCode();
		//if not in Dev mode, make sure the message and code is masked for unanticipated exceptions
		if (!$isDevMode) {
			/** Log information about actual exception */
			$reportId = $this->_critical($exception);
			$message = sprintf(self::INTERNAL_SERVER_ERROR_MSG, $reportId);
			$code = 0;
		}
		$maskedException = new WebapiException(
			new Phrase($message),
			$code,
			WebapiException::HTTP_INTERNAL_ERROR,
			[],
			'',
			null,
			$stackTrace
		);
	}
	return $maskedException;
}

```


As I am combing the situation, I am finding that it is too complicated to deal with in one section of code. The Exception classes all inherit a $code variable, and it would be better practice to set that for handlers to know what the proper HTTP response should be in situations.

400 user error or missing input
401 not authorized
403 never authorized (best to do 404)
404 not found (presently handled)
409 Conflict, that the user can correct (the payment information or authorization didn't match, try again is an acceptable situation)
410 Gone, the resource is no longer here (expired, InvalidState, don't expect this to work in the future)
500 General, non distinct error
502 Bad response from a remote server (gateway for 3rd party services such as card processing)
503 This service is unavailable or down. 
504 The remote server timed out (gateway for 3rd party services such as card processing)

I am contacting other developers (that we use) who make modules that do not pass the $code back

**Comment by magento-engcom-team** [comment]
Hi @ppostma1. 

Thank you for your report and collaboration!

The related internal Jira ticket `MAGETWO-97817` was closed as `non-reproducible` in `2.2.x`.
It means that Magento team either unable to reproduce this issue using provided _Steps to Reproduce_ from the **_Description section_** on clean or the issue has been already fixed in the scope of other tasks.

But if you still run into this problem please update or provide additional information/steps/preconditions in the **_Description section_** and reopen this issue.

**Comment by andrewbess** [comment]
Hi. I tried to reproduce it and I cannot reproduce it
I unassign this issue from me.
